gnome-shell: Fix visual glitches when Arc theme is applied to GDM lock/login screens

drakkar1969 opened the pull request:
In addition to fixing some typos in the gnome-shell theme (carry-overs from horst's original theme), this PR fixes a few visual glitches that only appear when applying the Arc gnome-shell theme to the GDM login/lock screens (by replacing the default theme located at `/usr/share/gnome-shell/gnome-shell-theme.gresource`, see https://github.com/NicoHood/arc-theme/issues/98).
Essentially this adds some css properties from the default shell theme, that are normally "pulled in" from the default `.gresource` file; if however the default `.gresource` file is replaced with a custom Arc one, these default properties are lost, so to speak.

**Screenshots of visual glitches when using a custom Arc `.gresource` file to style GDM**:

- Missing right border on modal dialog buttons:

   ![modal-dialog-button](https://user-images.githubusercontent.com/18715287/59754161-dbf3bb80-9285-11e9-8190-e0a6b7537b2e.png)

- Reduced/incorrect margin between run dialog entry and button:

   ![run-dialog-entry](https://user-images.githubusercontent.com/18715287/59754245-02195b80-9286-11e9-9a31-9ec119952d15.png)

- Background of last item in popup menu sub-menu overlapping sub-menu bottom border/box-shadow:

   ![popup-menu-sub-menu](https://user-images.githubusercontent.com/18715287/59754367-3ee55280-9286-11e9-8350-fac95a429b6b.png)

- Missing border for boxes in workspace switcher:

   ![workspace-switcher](https://user-images.githubusercontent.com/18715287/59754411-56bcd680-9286-11e9-94c5-7afb796ccdea.png)

- Missing border around dash (dash-to-dock not affected):

   ![dash-border](https://user-images.githubusercontent.com/18715287/59754450-6a683d00-9286-11e9-9e09-26b67562bd4a.png)

- Mis-aligned show apps button in dash (and dash-to-dock):

   ![dash-show-apps](https://user-images.githubusercontent.com/18715287/59754508-853ab180-9286-11e9-8b83-c29f1ddfbee7.png)

- Missing arrow in ibus candidate popup:

   ![ibus-candidate](https://user-images.githubusercontent.com/18715287/59754600-b1eec900-9286-11e9-98bd-ca298544e018.png)

- Missing arrow in overview collection (app folder) popup:

   ![overview-folder](https://user-images.githubusercontent.com/18715287/59754636-c4690280-9286-11e9-8bf0-bcc3c2f0d105.png)

jnsh commented:
Thank you. This is very nice work.

Just to make sure, did you test, or check the upstream shell-theme code, to make sure the the updates for the `.gresource` are correct all the way back to the 3.18 shell version?

Few small notes about the commit messages:
- I think changing the `when Arc theme is used to theme GDM` to e.g. `when Arc theme is compiled into a gresource file` would be more descriptive, even though GDM theming is the most common use-case for this.
- The subject line of the commit messages should be prepended with `gnome-shell: `.
- The message body should be wrapped around at 70 characters for github. For https://github.com/jnsh/arc-theme/pull/3/commits/00d98b8ae1da90eb25479e7e86d95c7652440833 and https://github.com/jnsh/arc-theme/pull/3/commits/da24a49fc2833238827ec4788706ac7a458454d1 the commit message wraps unevenly in the commit list.

(I can also fix the above myself before merging.)

Do you happen to have the `gresource.xml` file around to share so I could test this myself? Eventually it would be nice to have at least the `.xml` file generated during the build, but I haven't looked at that yet due to more important issues. This PR is very nice groundwork ahead of that.

I have large update to gnome-shell theme on the works, and nearly finished. Few of these commits may conflict slightly with that work, and I need to consider whether I should push that before merging this. I'm hoping to get that finished shortly, and then have a final look.

jnsh commented on common/gnome-shell/3.18/sass/_common.scss:
```diff
@@ -125,7 +125,7 @@ StScrollBar {
   StButton#vhandle, StButton#hhandle {
     border-radius: 4px;
     background-color: mix($fg_color, $bg_color, 40%);
-    border: 0px solid;
```
Technically this is not a typo, or in any way incorrect, although defining the `solid` border-style seems completely redundant for 0px wide border. At least the commit message could be changed to reflect that.

I'm wondering though, whether the border-style could have been defined on purpose due to some gnome-shell CSS oddity. Just wondering if you have investigated this any furter? If you tested that the scrollbar handles look as they should after this, I suppose this should be ok for merging.

drakkar1969 replied:
You're correct that this is not technically a typo, the redundant border style was just bugging me - I will amend the commit message to be more accurate.

I have tested this visually, and scrollbar handles are unchanged with this edit.

The upstream code actually doesn't have a border property for the scrollbar handles (there is a commented out border property), so this was added by horst for the Arc theme.

jnsh replied:
I think it should be ok then, to remove the border definition altogether, if that doesn't change the appearance.

I'm not sure I understand the purpose of the commented out border definition in upstream, as the explanation, `//would be nice to margin or at least to transparent` makes little sense to me. My best guess is that the 3px margin could be replaced with 3px transparent border, which would allow the scrollbar handles to be click-and-dragged from the margin area.

If upstream adds the border at some point, it can be then updated for Arc, but for now the border definition seems completely redundant to me.

drakkar1969 replied:
After some testing, removing the `border` property completely does not alter the visual appearance of the scrollbar handles; have tested both with and without the Arc theme compiled into the `gresource` file.

So I do agree that the best option is to remove the `border` property completely, seeing as it is absent in upstream.

jnsh commented on common/gnome-shell/3.18/sass/_common.scss:
```diff
@@ -687,8 +687,8 @@ StScrollBar {
-  -arrow-rise: 5;
+  -arrow-base: 20px;
+  -arrow-rise: 10px;
 }
```
With this, and https://github.com/jnsh/arc-theme/pull/3/commits/cf0f8eb5ca74ad3277ccd231ec85c4730343de64, I'm wondering what the original design looked like on gnome-shell 3.18, when the `-arrow-base` and `-arrow-rise` definitions were added. The popup menu arrows aren't used in other gnome-shell's popup-menus, and I'm wondering if that was the intention here as well.

I don't have GNOME 3.18 set up on a VM for testing so I can't check right now. Just wondering if you investigated this further. Also what did you use as a basis for the new pixel values?

drakkar1969 replied:
The `-arrow-base` and `arrow-rise` properties are present in horst's theme as far back as the initial commits for version 3.14, so it's hard to see what the design change was.

Since assigning a non-zero numerical value without units to these properties is an error/typo, I was trying to figure out what the original design intent was. The `-arrow-base` and `arrow-rise` properties appear 4 times in the Arc gnome shell theme:
- `.popup-menu-boxpointer`: the properties are zeroed out, so obviously the intent was to hide the arrow
- `.keyboard-subkeys`: explicit values assigned (20px base, 10px rise) to ensure arrow is displayed
- `.candidate-popup-boxpointer` and `.app-folder-popup`: assigning a value of 5 to the properties is an error, so currently the Arc theme falls back on the values in the default shell theme (in both cases `-arrow-base: 24px` and `-arrow-rise: 11px`) and the arrow ends up being displayed in both cases. Since horst then modified some of the other `-arrow-*` properties in `.candidate-popup-boxpointer` and `.app-folder-popup`, I went with the assumption that the intent was to have the arrow displayed in both cases. The pixel values I used (20px/10px) are consistent with those used by the Arc theme in `.keyboard-subkeys`, for uniformity.

jnsh replied:
Thanks for the clarification: I'm happy to hear you considered this thoroughly :)

I think it's safe to assume that the original intention was to have the arrow-pointers visible, and this would definitely make sense at least for the `.app-folder-popup`, since without the arrow there's no way to determine which app-folder is opened if there are multiple folders side-by-side. Re-using the values from `.keyboard-subkeys` also makes sense.

drakkar1969 commented:
> Just to make sure, did you test, or check the upstream shell-theme code, to make sure the the updates for the .gresource are correct all the way back to the 3.18 shell version?

I have checked the upstream code back to version 3.18 to ensure that properties were present. Only exception is on commit https://github.com/jnsh/arc-theme/pull/3/commits/35d6eab9febcc019cfd0a390cfcd391e8163058e where I have added a couple of border width properties for rtl mode:

```
&:rtl {
    border-left-width: 1px;
    border-right-width: 0;
    border-radius: 3px 0 0 3px;
}
```

I don't see how else the dash border would be correctly rendered in rtl mode: if the border radius is flipped when  switching from ltr to rtl, so should the left/right border widths.

> Few small notes about the commit messages:
> 
> * I think changing the `when Arc theme is used to theme GDM` to e.g. `when Arc theme is compiled into a gresource file` would be more descriptive, even though GDM theming is the most common use-case for this.
> * The subject line of the commit messages should be prepended with `gnome-shell: `.
> * The message body should be wrapped around at 70 characters for github. For [00d98b8](https://github.com/jnsh/arc-theme/commit/00d98b8ae1da90eb25479e7e86d95c7652440833) and [da24a49](https://github.com/jnsh/arc-theme/commit/da24a49fc2833238827ec4788706ac7a458454d1) the commit message wraps unevenly in the commit list.

Fair points, happy to update the commit messages. I can do that as soon as the various conversations here are resolved (only waiting because I think this will change the commit hashes, and not sure if github will keep the cross-references between conversations and commits if hashes change).

> Do you happen to have the `gresource.xml` file around to share so I could test this myself?

These are the xml files I use:

**Arc**:
```
<?xml version="1.0" encoding="UTF-8"?>
<gresources>
  <gresource prefix="/org/gnome/shell/theme">
    <file>common-assets/dash/dash-left.svg</file>
    <file>common-assets/dash/dash-placeholder.svg</file>
    <file>common-assets/dash/dash-right.svg</file>
    <file>common-assets/dash/running1.svg</file>
    <file>common-assets/dash/running2.svg</file>
    <file>common-assets/dash/running3.svg</file>
    <file>common-assets/dash/running4.svg</file>
    <file>common-assets/key/key-enter.svg</file>
    <file>common-assets/key/key-hide.svg</file>
    <file>common-assets/key/key-layout.svg</file>
    <file>common-assets/key/key-shift-latched-uppercase.svg</file>
    <file>common-assets/key/key-shift.svg</file>
    <file>common-assets/key/key-shift-uppercase.svg</file>
    <file>common-assets/menu/menu-hover.svg</file>
    <file>common-assets/menu/menu-separator.svg</file>
    <file>common-assets/misc/bg.svg</file>
    <file>common-assets/misc/calendar-today.svg</file>
    <file>common-assets/misc/more-results.svg</file>
    <file>common-assets/misc/osd.svg</file>
    <file>common-assets/misc/process-working.svg</file>
    <file>common-assets/misc/ws-switch-arrow-down.png</file>
    <file>common-assets/misc/ws-switch-arrow-up.png</file>
    <file>common-assets/panel/activities-active.svg</file>
    <file>common-assets/panel/activities.svg</file>
    <file>common-assets/panel/panel-overview.svg</file>
    <file>common-assets/panel/panel.svg</file>
    <file>common-assets/switch/switch-off-selected.svg</file>
    <file>common-assets/switch/switch-on-selected.svg</file>
    <file>gnome-shell.css</file>
    <file>light-assets/checkbox/checkbox-checked-focused.svg</file>
    <file>light-assets/checkbox/checkbox-checked.svg</file>
    <file>light-assets/checkbox/checkbox-unchecked-focused.svg</file>
    <file>light-assets/checkbox/checkbox-unchecked.svg</file>
    <file>light-assets/menu/menu.svg</file>
    <file>light-assets/menu/submenu-open.svg</file>
    <file>light-assets/menu/submenu.svg</file>
    <file>light-assets/misc/calendar-arrow-left-hover.svg</file>
    <file>light-assets/misc/calendar-arrow-left.svg</file>
    <file>light-assets/misc/calendar-arrow-right-hover.svg</file>
    <file>light-assets/misc/calendar-arrow-right.svg</file>
    <file>light-assets/misc/message-active.svg</file>
    <file>light-assets/misc/message-close-active.svg</file>
    <file>light-assets/misc/message-close-hover.svg</file>
    <file>light-assets/misc/message-close.svg</file>
    <file>light-assets/misc/message-hover.svg</file>
    <file>light-assets/misc/message.svg</file>
    <file>light-assets/misc/modal.svg</file>
    <file>light-assets/switch/switch-off.svg</file>
    <file>light-assets/switch/switch-on.svg</file>
  </gresource>
</gresources>
```

**Arc-Dark**:
```
<?xml version="1.0" encoding="UTF-8"?>
<gresources>
  <gresource prefix="/org/gnome/shell/theme">
    <file>common-assets/dash/dash-left.svg</file>
    <file>common-assets/dash/dash-placeholder.svg</file>
    <file>common-assets/dash/dash-right.svg</file>
    <file>common-assets/dash/running1.svg</file>
    <file>common-assets/dash/running2.svg</file>
    <file>common-assets/dash/running3.svg</file>
    <file>common-assets/dash/running4.svg</file>
    <file>common-assets/key/key-enter.svg</file>
    <file>common-assets/key/key-hide.svg</file>
    <file>common-assets/key/key-layout.svg</file>
    <file>common-assets/key/key-shift-latched-uppercase.svg</file>
    <file>common-assets/key/key-shift.svg</file>
    <file>common-assets/key/key-shift-uppercase.svg</file>
    <file>common-assets/menu/menu-hover.svg</file>
    <file>common-assets/menu/menu-separator.svg</file>
    <file>common-assets/misc/bg.svg</file>
    <file>common-assets/misc/calendar-today.svg</file>
    <file>common-assets/misc/more-results.svg</file>
    <file>common-assets/misc/osd.svg</file>
    <file>common-assets/misc/process-working.svg</file>
    <file>common-assets/misc/ws-switch-arrow-down.png</file>
    <file>common-assets/misc/ws-switch-arrow-up.png</file>
    <file>common-assets/panel/activities-active.svg</file>
    <file>common-assets/panel/activities.svg</file>
    <file>common-assets/panel/panel-overview.svg</file>
    <file>common-assets/panel/panel.svg</file>
    <file>common-assets/switch/switch-off-selected.svg</file>
    <file>common-assets/switch/switch-on-selected.svg</file>
    <file>dark-assets/checkbox/checkbox-checked-focused.svg</file>
    <file>dark-assets/checkbox/checkbox-checked.svg</file>
    <file>dark-assets/checkbox/checkbox-unchecked-focused.svg</file>
    <file>dark-assets/checkbox/checkbox-unchecked.svg</file>
    <file>dark-assets/menu/menu.svg</file>
    <file>dark-assets/menu/submenu-open.svg</file>
    <file>dark-assets/menu/submenu.svg</file>
    <file>dark-assets/misc/calendar-arrow-left-hover.svg</file>
    <file>dark-assets/misc/calendar-arrow-left.svg</file>
    <file>dark-assets/misc/calendar-arrow-right-hover.svg</file>
    <file>dark-assets/misc/calendar-arrow-right.svg</file>
    <file>dark-assets/misc/message-active.svg</file>
    <file>dark-assets/misc/message-close-active.svg</file>
    <file>dark-assets/misc/message-close-hover.svg</file>
    <file>dark-assets/misc/message-close.svg</file>
    <file>dark-assets/misc/message-hover.svg</file>
    <file>dark-assets/misc/message.svg</file>
    <file>dark-assets/misc/modal.svg</file>
    <file>dark-assets/switch/switch-off.svg</file>
    <file>dark-assets/switch/switch-on.svg</file>
    <file>gnome-shell.css</file>
  </gresource>
</gresources>
```

jnsh commented:
> Only exception is on commit [35d6eab](https://github.com/jnsh/arc-theme/commit/35d6eab9febcc019cfd0a390cfcd391e8163058e) where I have added a couple of border width properties for rtl mode:
> 
> ```
> &:rtl {
>     border-left-width: 1px;
>     border-right-width: 0;
>     border-radius: 3px 0 0 3px;
> }
> ```
> 
> I don't see how else the dash border would be correctly rendered in rtl mode: if the border radius is flipped when switching from ltr to rtl, so should the left/right border widths.

I did notice this earlier, but didn't comment on it as this indeed seems correct for the `:rtl`.

However it seems a bit odd that this would have left unnoticed for upstream though, and I'm thinking there's a chance that the border's are flipped around automatically for `:rtl` somewhere in the UI JavaScript code. It's probably a good idea to make sure before merging.

I just need to figure out how to set up a right-to-left locale to test this (I don't think there's any other way to test), and there are some other `:rtl` related things I need to test anyway.

If the border changes for `:rtl` are necessary and correct, this should also be fixed for the upstream theme. 

> > Few small notes about the commit messages:

One more small thing: Maybe the typo-fix commits could be merged to a single commit? They all share the same purpose, and there's very small chance that any of them would have to be reverted. This would make for cleaner commit log, but it's not a major issue, and I leave this up to your consideration.

drakkar1969 commented:
> > Only exception is on commit [35d6eab](https://github.com/jnsh/arc-theme/commit/35d6eab9febcc019cfd0a390cfcd391e8163058e) where I have added a couple of border width properties for rtl mode:
> > ```
> > &:rtl {
> >     border-left-width: 1px;
> >     border-right-width: 0;
> >     border-radius: 3px 0 0 3px;
> > }
> > ```
> > 
> > 
> > I don't see how else the dash border would be correctly rendered in rtl mode: if the border radius is flipped when switching from ltr to rtl, so should the left/right border widths.
> 
> I did notice this earlier, but didn't comment on it as this indeed seems correct for the `:rtl`.
> 
> However it seems a bit odd that this would have left unnoticed for upstream though, and I'm thinking there's a chance that the border's are flipped around automatically for `:rtl` somewhere in the UI JavaScript code. It's probably a good idea to make sure before merging.
> 
> I just need to figure out how to set up a right-to-left locale to test this (I don't think there's any other way to test), and there are some other `:rtl` related things I need to test anyway.
> 
> If the border changes for `:rtl` are necessary and correct, this should also be fixed for the upstream theme.

I did some testing on a VM with rtl locale, and this does seem to be a bug/omission in upstream - the dash border does not get switched around automatically in `:rtl` mode.

Screenshots (with dash border color changed to yellow for better visibility):

Without extra `border-left-width: 1px` and `border-right-width: 0` properties in `:rtl`:
![dash-rtl-orig](https://user-images.githubusercontent.com/18715287/59922978-6d079580-9432-11e9-8745-af9f225648ed.png)

With the extra properties in `:rtl` as per commit https://github.com/jnsh/arc-theme/pull/3/commits/35d6eab9febcc019cfd0a390cfcd391e8163058e:
![dash-rtl-mod](https://user-images.githubusercontent.com/18715287/59923040-a8a25f80-9432-11e9-9a38-973dcd34e07e.png)

drakkar1969 commented:
Based on all of the above, I have made the following to changes to my local copy:
* Prepended all commit messages with `gnome-shell: `
* Changed commit descriptions to include `when Arc theme is compiled into a gresource file`
* Squashed the `Fix typo` commits (https://github.com/jnsh/arc-theme/pull/3/commits/fe7ea8fa4adfe99faeca43b0825740e1c9114860, https://github.com/jnsh/arc-theme/pull/3/commits/617d53b50ea9895260911a402b6b6ad67de69784 and https://github.com/jnsh/arc-theme/pull/3/commits/8d20e78a13d33089c7298acb198fc2f941c97ac9) into a single commit
* Modified the `Fix typo in scrollbar border` commit (https://github.com/jnsh/arc-theme/pull/3/commits/c827aa2020001650032748a0f16a7cbcd523b5d7) to completely remove the `border` property for scrollbar handles (and modified commit message accordingly)
* Left the content of the `Add explicit border to dash` commit (https://github.com/jnsh/arc-theme/pull/3/commits/35d6eab9febcc019cfd0a390cfcd391e8163058e) unchanged, as per my previous comment

Happy to push changes if and when you are ok with this/all conversation are resolved

jnsh commented:
> I did some testing on a VM with rtl locale, and this does seem to be a bug/omission in upstream - the dash border does not get switched around automatically in :rtl mode.

Thank you for testing this. I'd imagine gnome-shell devs wouldn't mind having this fixed in the default shell theme as well, if you're interested in contributing to upstream.

I think everything should be in order after you push the changes, and I'm happy to merge this after I give it a quick test run.

jnsh commented:
Pushed to `master`.
Thank you again.

jnsh commented:
I already merged and pushed this once, but decided to have a rethink of how I want to perform pull request merging on this repo. I will merge again once I figure it out.

I tested the changes with gresource and everything looks good. Thank you again for this work.

drakkar1969 commented:
> > I did some testing on a VM with rtl locale, and this does seem to be a bug/omission in upstream - the dash border does not get switched around automatically in :rtl mode.
> 
> Thank you for testing this. I'd imagine gnome-shell devs wouldn't mind having this fixed in the default shell theme as well, if you're interested in contributing to upstream.

Opened a merge request upstream: https://gitlab.gnome.org/GNOME/gnome-shell/merge_requests/589

> I already merged and pushed this once, but decided to have a rethink of how I want to perform pull request merging on this repo. I will merge again once I figure it out, probably within a few days time. Sorry about the mess.

No worries :)

jnsh commented:
Merged with the above commits.
Apparently there's no way to mark the PR as "Merged" when pulling via command line, so I'm closing this as merged.

Thank you once again.
